### ISSUE TITLE
[054] Freeze a Python reference and asset-conversion environment

### ISSUE THREAD

**Issue by 25d-228** [25d-228/gpt-2-C#55]
**Prerequisite:** #54 and #1. **Budget:** 30–45 minutes of setup/explanation; package downloads are separate. Python is an independent oracle and converter, never the deployed model runtime.

## Scope and fixed choices
Create `requirements-reference.txt`, `tools/check_reference_env.py`, `tests/test_reference_env.py`, and `docs/reference-environment.md`. Use Python3.11 in `.venv` and these direct pins: `torch==2.6.0`, `transformers==4.51.3`, `numpy==2.2.6`, `safetensors==0.5.3`, `regex==2024.11.6`, `huggingface-hub==0.30.2`. These are deliberately selected reference versions, not a claim to be the latest. Use the CPU PyTorch wheel index on Linux/WSL; use the matching supported wheel on macOS. Keep the environment isolated from preexisting packages. After installation save `pip freeze` and platform/Python details in a platform-specific reference lock and use that lock for reruns; do not silently upgrade dependencies to fix a mismatch. A deliberate dependency/security update requires re-running the reference gates and recording a new lock, not pretending it is the same environment.

Read installed package versions through `importlib.metadata.version` using distribution names. Compare the base release version, allowing torch's expected local CPU suffix (for example `2.6.0+cpu`); record the full version including that suffix. Do not compare regex's internal `__version__` string to its date-based distribution version. The expected regex distribution version is2024.11.6. Test these normalization rules without changing installed packages.

The AI may provide and explain these Python helpers. Construct an offline `GPT2LMHeadModel` with V7,maxT8,L2,H2,C4, gelu_new, epsilon1e-5, all dropout0, eager attention, CPU, eval mode and use_cache=False. Use inputs `[[0,1,2],[2,1,0]]`; verify42 finite logits. Compute finite loss with explicit functional cross-entropy against already-shifted targets `[[1,2,3],[1,0,3]]`, as #1 specifies, rather than using a second internal target shift. No pretrained download yet. Add a tiny safetensors save/read round-trip and regex Unicode-property smoke test.

## Exact procedure
```
python3.11 -m venv .venv
.venv/bin/python -m pip install torch==2.6.0 --index-url https://download.pytorch.org/whl/cpu
.venv/bin/python -m pip install -r requirements-reference.txt
.venv/bin/python -m pip check
.venv/bin/python -m unittest discover -s tests -p 'test_reference_env.py'
.venv/bin/python tools/check_reference_env.py
make test
```
The CPU-index line is Linux/WSL-specific; on macOS install torch==2.6.0 from the ordinary supported wheel index, then the same requirements. Tests assert versions, offline construction, shape(2,3,7), six contributing loss rows, tensor equality after safe serialization, Unicode-letter matching, and rejection of a deliberately wrong expected release. Set torch threads1 and record deterministic settings. Explicitly test that version2.6.0+cpu passes the2.6.0 release check and2.7.0 fails. Dependency availability is not proof that C parity passes.

**Mutation:** leave the model in training mode with nonzero dropout; the helper must flag the mode/config before approving reference output. Restore eval settings. Do not rely on chance output differences alone to detect a wrong mode.

**Exit question:** Why is using PyTorch to check your C implementation different from delegating your runtime to PyTorch?

References: https://huggingface.co/docs/transformers/v4.51.3/model_doc/gpt2 and version-specific PyPI pages. Record exact environment and results in `notes/step-054.md`; follow #1.
